android/java/org/chromium/chrome/browser/bookmarks/BraveBookmarkUtils.java

```diff
@@ -0,0 +1,153 @@
+            }
+            bookmarkModel.deleteBookmark(bookmarkId);
+            bookmarkModel.destroy();
+            return bookmarkId;
+        }
+
+        BookmarkId bookmarkId =
+                addBookmarkInternal(activity, bookmarkModel, tab.getTitle(), tab.getOriginalUrl());
+
+        Snackbar snackbar = null;
+        if (bookmarkId == null) {
+            snackbar = Snackbar.make(activity.getString(R.string.bookmark_page_failed),
+                                       new SnackbarController() {
+                                           @Override
+                                           public void onDismissNoAction(Object actionData) {}
+
+                                           @Override
+                                           public void onAction(Object actionData) {}
+                                       },
+                                       Snackbar.TYPE_NOTIFICATION, Snackbar.UMA_BOOKMARK_ADDED)
+                               .setSingleLine(false);
+            RecordUserAction.record("EnhancedBookmarks.AddingFailed");
+        } else {
+            String folderName = bookmarkModel.getBookmarkTitle(
+                    bookmarkModel.getBookmarkById(bookmarkId).getParentId());
+            SnackbarController snackbarController =
+                    createSnackbarControllerForEditButton(activity, bookmarkId);
+            if (BookmarkUtils.getLastUsedParent(activity) == null) {
+                if (fromCustomTab) {
+                    String packageLabel = BuildInfo.getInstance().hostPackageLabel;
+                    snackbar = Snackbar.make(
+                            activity.getString(R.string.bookmark_page_saved, packageLabel),
+                            snackbarController, Snackbar.TYPE_ACTION, Snackbar.UMA_BOOKMARK_ADDED);
+                } else {
+                    snackbar = Snackbar.make(
+                            activity.getString(R.string.bookmark_page_saved_default),
+                            snackbarController, Snackbar.TYPE_ACTION, Snackbar.UMA_BOOKMARK_ADDED);
+                }
+            } else {
+                snackbar = Snackbar.make(folderName, snackbarController, Snackbar.TYPE_ACTION,
+                                           Snackbar.UMA_BOOKMARK_ADDED)
+                                   .setTemplateText(
+                                           activity.getString(R.string.bookmark_page_saved_folder));
+            }
+            snackbar.setSingleLine(false).setAction(
+                    activity.getString(R.string.bookmark_item_edit), null);
+        }
+        snackbarManager.showSnackbar(snackbar);
+
+        bookmarkModel.destroy();
+        return bookmarkId;
+    }
+
+    /**
+     * An internal version of {@link #addBookmarkSilently(Context, BookmarkModel, String, String)}.
+     * Will reset last used parent if it fails to add a bookmark
+     */
+    private static BookmarkId addBookmarkInternal(
```

Comment:
Created an issue to add support for static functions. https://github.com/brave/brave-browser/issues/13359 so instead of adding a patch, i have copied the function here.

Reply:
does that function somehow different of what we have inside BookmarkUtils.java? Why do we need to duplicate it as we can call it via that `BookmarkUtils.addBookmarkInternal` as it's static.